### PR TITLE
generate golden SEI

### PR DESCRIPTION
**Pull request by mithydolphin**
### Describe your changes

Please include a summary of the change, a relevant motivation and context.

### Issue ticket number and link

- Fixes #(issue)

### Checklist before requesting a review

- [ ] I have performed a self-review of my own code
- [ ] I have verified that the code builds perfectly fine on my local system
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have verified that my code follows the style already available in the repository
- [ ] I have made corresponding changes to the documentation

### CODE REVIEW

File: lib/src/signed_video_tlv.c [hunk 1]
```diff
@@ -1024,7 +1024,7 @@ tlv_list_encode_or_get_size(signed_video_t *self,
       continue;
     }
 
-    if (tlv.is_always_present || self->has_recurrent_data) {
+    if (tlv.is_always_present || self->has_recurrent_data || self->is_golden_sei) {
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:1027]: This is not the way to go. It is better if you set `self->has_recurrent_data = true;` in your new API. This recurrence stuff will be removed when we support golden SEI anyhow.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
Some initial review comments.

File: lib/src/signed_video_h26x_sign.c [hunk 2]
```diff
@@ -186,24 +186,26 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
   const size_t hash_size = sign_data->hash_size;
 
   // Metadata + hash_list forming a document.
-  const sv_tlv_tag_t document_encoders[] = {
+  const sv_tlv_tag_t golden_sei_document_encoders[] = {PUBLIC_KEY_TAG, PRODUCT_INFO_TAG};
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:189]: You also need the CRYPTO_INFO_TAG

File: lib/src/signed_video_h26x_sign.c [hunk 3]
```diff
@@ -186,24 +186,26 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
+  const sv_tlv_tag_t gop_info_encoders[] = {
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:190]: I would prefer to keep this declared last as before, since it by definition should end up last in the SEI.

File: lib/src/signed_video_h26x_sign.c [hunk 4]
```diff
@@ -186,24 +186,26 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
   const size_t hash_size = sign_data->hash_size;
 
   // Metadata + hash_list forming a document.
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:188]: This comment is not really true anymore. Rewrite.

File: lib/src/signed_video_h26x_sign.c [hunk 5]
```diff
@@ -186,24 +186,26 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
       GENERAL_TAG,
       CRYPTO_INFO_TAG,
       PUBLIC_KEY_TAG,
       PRODUCT_INFO_TAG,
       ARBITRARY_DATA_TAG,
       HASH_LIST_TAG,
   };
-  const sv_tlv_tag_t gop_info_encoders[] = {
-      SIGNATURE_TAG,
-  };
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:201]: May I suggest to declare one list of "always" encoders/tags `always_tags` and one including "only need once" encoders/tags `once_tags`.
This will make life easier when combining them in the different scenarios.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/signed_video_h26x_sign.c [hunk 6]
```diff
@@ -184,26 +184,23 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
 {
   sign_or_verify_data_t *sign_data = self->sign_data;
   const size_t hash_size = sign_data->hash_size;
+  size_t num_optional_document_encoders = 0;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:187]: I think you can shorten this to `num_optional_tags`

File: lib/src/signed_video_h26x_sign.c [hunk 7]
```diff
@@ -184,26 +184,23 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
+  size_t num_man_doc_encoders = 0;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:188]: I think you can shorten this to `num_mandatory_tags`

File: lib/src/signed_video_h26x_sign.c [hunk 8]
```diff
@@ -184,26 +184,23 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
-  // Metadata + hash_list forming a document.
-  const sv_tlv_tag_t document_encoders[] = {
-      GENERAL_TAG,
-      CRYPTO_INFO_TAG,
-      PUBLIC_KEY_TAG,
-      PRODUCT_INFO_TAG,
-      ARBITRARY_DATA_TAG,
-      HASH_LIST_TAG,
-  };
+  const sv_tlv_tag_t *optional_document_encoders =
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:190]: I think you can shorten this to `optional_tags`

File: lib/src/signed_video_h26x_sign.c [hunk 9]
```diff
@@ -184,26 +184,23 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
+      get_optional_tags(&num_optional_document_encoders);
+  const sv_tlv_tag_t *mandatory_document_encoders = get_mandatory_tags(&num_man_doc_encoders);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:192]: I think you can shorten this to `mandatory_tags`

File: lib/src/signed_video_h26x_sign.c [hunk 10]
```diff
@@ -224,21 +221,32 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
   SVI_TRY()
     // Get the total payload size of all TLVs. Then compute the total size of the SEI NALU to be
     // generated. Add extra space for potential emulation prevention bytes.
-    document_size = tlv_list_encode_or_get_size(self, document_encoders, num_doc_encoders, NULL);
+    optional_document_size = tlv_list_encode_or_get_size(
+        self, optional_document_encoders, num_optional_document_encoders, NULL);
+    mandatory_document_size = !self->is_golden_sei
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:226]: I think you should split this as
```
mandatory_document_size = tlv_list_encode_or_get_size(self, mandatory_document_encoders, num_man_doc_encoders, NULL);
if (self->is_golden_sei) {
  mandatory_document_size = 0;
}
```

File: lib/src/signed_video_h26x_sign.c [hunk 11]
```diff
@@ -224,21 +221,32 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
+    document_size = optional_document_size + mandatory_document_size;
     payload_size = document_size + gop_info_size + vendor_size;
     payload_size += UUID_LEN;  // UUID
     payload_size += 1;  // One byte for reserved data.
     if ((self->max_sei_payload_size > 0) && (payload_size > self->max_sei_payload_size)) {
       // Fallback to GOP-level signing
       payload_size -= document_size;
       self->gop_info->list_idx = -1;  // Reset hash list size to exclude it from TLV
-      document_size = tlv_list_encode_or_get_size(self, document_encoders, num_doc_encoders, NULL);
+      optional_document_size = tlv_list_encode_or_get_size(
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:243]: Remove this line. Removing the hash list will not change the optional tags. It is part of the mandatory tags.

File: lib/src/signed_video_h26x_sign.c [hunk 12]
```diff
@@ -224,21 +221,32 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
     if (self->num_vendor_encoders > 0 && self->vendor_encoders) {
       vendor_size =
           tlv_list_encode_or_get_size(self, self->vendor_encoders, self->num_vendor_encoders, NULL);
     }
 
+    document_size = optional_document_size + mandatory_document_size;
     payload_size = document_size + gop_info_size + vendor_size;
     payload_size += UUID_LEN;  // UUID
     payload_size += 1;  // One byte for reserved data.
     if ((self->max_sei_payload_size > 0) && (payload_size > self->max_sei_payload_size)) {
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:239]: Add a `&& mandatory_document_size > 0` in the if-statement, since it only affects the mandatory part. This way you do not need tho check the `is_golden_sei` flag inside either.

File: lib/src/signed_video_h26x_sign.c [hunk 13]
```diff
@@ -224,21 +221,32 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
       // Fallback to GOP-level signing
       payload_size -= document_size;
       self->gop_info->list_idx = -1;  // Reset hash list size to exclude it from TLV
-      document_size = tlv_list_encode_or_get_size(self, document_encoders, num_doc_encoders, NULL);
+      optional_document_size = tlv_list_encode_or_get_size(
+          self, optional_document_encoders, num_optional_document_encoders, NULL);
+      mandatory_document_size = !self->is_golden_sei
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:245]: Replace with `mandatory_document_size = tlv_list_encode_or_get_size(self, mandatory_document_encoders, num_man_doc_encoders, NULL);`
if you follow my advice to add the mandatory size in the if above.

File: lib/src/signed_video_h26x_sign.c [hunk 14]
```diff
@@ -296,10 +304,16 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
     reserved_byte |= self->is_golden_sei << 6;
     *payload_ptr++ = reserved_byte;
 
-    size_t written_size =
-        tlv_list_encode_or_get_size(self, document_encoders, num_doc_encoders, payload_ptr);
+    size_t optional_written_size = tlv_list_encode_or_get_size(
+        self, optional_document_encoders, num_optional_document_encoders, payload_ptr);
+    payload_ptr += optional_written_size;
+    size_t mandatory_written_size = !self->is_golden_sei
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:310]: I would prefer something like this instead
```
size_t mandatory_written_size = 0;
if (mandatory_document_size > 0) {
  mandatory_written_size = tlv_list_encode_or_get_size(self, mandatory_document_encoders, num_man_doc_encoders, payload_ptr);
  payload_ptr += mandatory_written_size;
}
```

File: lib/src/signed_video_h26x_sign.c [hunk 15]
```diff
@@ -676,6 +690,37 @@ signed_video_set_end_of_stream(signed_video_t *self)
...
+    while (sv_signing_plugin_get_signature(self->plugin_handle, sign_data->signature,
+        sign_data->max_signature_size, &sign_data->signature_size, &signature_error)) {
+      SVI_THROW(sv_rc_to_svi_rc(signature_error));
+      SVI_THROW(complete_sei_nalu_and_add_to_prepend(self));
+    }
+
+  SVI_CATCH()
+  SVI_DONE(status)
+  self->is_golden_sei = false;
+  self->has_recurrent_data = false;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:720]: Remove. This flag is reset when it has been used.

File: lib/src/signed_video_h26x_sign.c [hunk 16]
```diff
@@ -676,6 +690,37 @@ signed_video_set_end_of_stream(signed_video_t *self)
...
+  self->is_golden_sei = true;
+  self->has_recurrent_data = true;
+  SVI_TRY()
+    SVI_THROW(prepare_for_nalus_to_prepend(self));
+    SVI_THROW(generate_sei_nalu(self, &payload, &payload_signature_ptr));
+    add_payload_to_buffer(self, payload, payload_signature_ptr);
+    // Fetch the signature. If it is not ready we exit without generating the SEI.
+    sign_or_verify_data_t *sign_data = self->sign_data;
+    SignedVideoReturnCode signature_error = SV_UNKNOWN_FAILURE;
+    // Note: From here, this is a temporary solution. It will only work unthreaded.
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:710]: This line should go right after `add_payload_to_buffer`

File: lib/src/signed_video_h26x_sign.c [hunk 17]
```diff
@@ -676,6 +690,37 @@ signed_video_set_end_of_stream(signed_video_t *self)
...
+  uint8_t *payload = NULL;
+  uint8_t *payload_signature_ptr = NULL;
+  svi_rc status = SVI_UNKNOWN;
+  self->is_golden_sei = true;
+  self->has_recurrent_data = true;
+  SVI_TRY()
+    SVI_THROW(prepare_for_nalus_to_prepend(self));
+    SVI_THROW(generate_sei_nalu(self, &payload, &payload_signature_ptr));
+    add_payload_to_buffer(self, payload, payload_signature_ptr);
+    // Fetch the signature. If it is not ready we exit without generating the SEI.
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:707]: Add an empty line right after `add_payload_to_buffer` to mark a new section.

File: lib/src/signed_video_h26x_sign.c [hunk 18]
```diff
@@ -676,6 +690,37 @@ signed_video_set_end_of_stream(signed_video_t *self)
...
 }
 
+SignedVideoReturnCode
+signed_video_generate_golden_sei(signed_video_t *self)
+{
+  if (!self) return SV_INVALID_PARAMETER;
+
+  uint8_t *payload = NULL;
+  uint8_t *payload_signature_ptr = NULL;
+  svi_rc status = SVI_UNKNOWN;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:700]: Put this just before `SVI_TRY()` since it belongs to the try-catch scope. Also add an empty line before it to mark a clear separation.

File: lib/src/signed_video_tlv.c [hunk 19]
```diff
@@ -132,6 +132,25 @@ static const sv_tlv_tuple_t tlv_tuples[] = {
...
+ * It is necessary to generate golden SEIs and when the golden SEI principle is not used.
+ */
+static const sv_tlv_tag_t optional_tags[] = {
+    CRYPTO_INFO_TAG,
+    PUBLIC_KEY_TAG,
+    PRODUCT_INFO_TAG,
+};
+
+/*
+ * This is an array that contains only mandatory tags (|is_always_present|).
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:146]: Add that it exclude the SIGNATURE_TAG since it has to be treated separately.

File: lib/src/signed_video_tlv.c [hunk 20]
```diff
@@ -132,6 +132,25 @@ static const sv_tlv_tuple_t tlv_tuples[] = {
     {NUMBER_OF_TLV_TAGS, 0, NULL, NULL, true},
 };
 
+/*
+ * This is an array that contains only optional tags (not |is_always_present|).
+ * It is necessary to generate golden SEIs and when the golden SEI principle is not used.
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:137]: You can leave out this line. If we see `signed_video_tlv` as a file that simply handles the TLV structure without asking any questions, it is of no interest to know how or when it is used. This comment belongs in the `sign` file.

File: lib/src/signed_video_tlv.c [hunk 21]
```diff
@@ -1160,6 +1179,19 @@ tlv_find_and_decode_recurrent_tags(signed_video_t *self,
   return recurrent_tags_decoded;
 }
 
+const sv_tlv_tag_t *
+get_optional_tags(size_t *num_of_op_tags)
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:1183]: You can afford `num_of_optional_tags`

File: lib/src/signed_video_tlv.c [hunk 22]
```diff
@@ -1160,6 +1179,19 @@ tlv_find_and_decode_recurrent_tags(signed_video_t *self,
...
+{
+  *num_of_op_tags = ARRAY_SIZE(optional_tags);
+  return optional_tags;
+}
+
+const sv_tlv_tag_t *
+get_mandatory_tags(size_t *num_of_man_tags)
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:1190]: You can afford `num_of_mandatory_tags`

File: lib/src/signed_video_tlv.c [hunk 23]
```diff
@@ -1160,6 +1179,19 @@ tlv_find_and_decode_recurrent_tags(signed_video_t *self,
...
+{
+  *num_of_man_tags = ARRAY_SIZE(mandatory_tags);
+  return mandatory_tags;
+}
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.c:1194]: Add an empty line between the functions.

File: lib/src/signed_video_tlv.h [hunk 24]
```diff
@@ -149,4 +149,24 @@ tlv_find_and_decode_recurrent_tags(signed_video_t *self,
     const uint8_t *tlv_data,
     size_t tlv_data_size);
 
+/**
+ * @brief Helper to get only the optional tags as an array.
+ *
+ * @param num_of_man_tags Number of optional tags.
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.h:155]: Rephrase to "A pointer to a location where the number of optional tags will be written."

File: lib/src/signed_video_tlv.h [hunk 25]
```diff
@@ -149,4 +149,24 @@ tlv_find_and_decode_recurrent_tags(signed_video_t *self,
...
+ *
+ * @returns Array that contains optional tags.
+ */
+const sv_tlv_tag_t *
+get_optional_tags(size_t *num_of_op_tags);
+
+/**
+ * @brief Helper to get only the mandatory tags as an array.
+
+ * @param num_of_man_tags Number of mandatory tags.
```
Comment (bjornvolcker) [inline lib/src/signed_video_tlv.h:165]: Rephrase to "A pointer to a location where the number of mandatory tags will be written."

File: tests/check/check_h26xsigned_sign.c [hunk 26]
```diff
@@ -673,6 +673,49 @@ START_TEST(two_completed_seis_pending)
 }
 END_TEST
 
+/* Test description
+ * Verify that after golden SEI is generated
+ * The operation is as follows:
+ * 1. Setup a signed_video_t session
+ * 2. Generate a golden SEI
+ * 3. Verify the golden SEI
+ */
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:682]: I think you can compress the test description to something like
"Generates a golden SEI and fetches it from the library. Then verifies that the correspond flag is set."

It could be a good idea to check that the PUBLIC_KEY tag is present as well, if so, add that info to the description.

File: tests/check/check_h26xsigned_sign.c [hunk 27]
```diff
@@ -673,6 +673,49 @@ START_TEST(two_completed_seis_pending)
...
+START_TEST(golden_sei_created)
+{
+
+  SignedVideoCodec codec = settings[_i].codec;
+  SignedVideoReturnCode sv_rc;
+  size_t sei_size = 0;
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:688]: May I suggest moving this declaration to where it is actually used.

File: tests/check/check_h26xsigned_sign.c [hunk 28]
```diff
@@ -673,6 +673,49 @@ START_TEST(two_completed_seis_pending)
...
+  ck_assert(sv);
+  char *private_key = NULL;
+  size_t private_key_size = 0;
+  // Setup the key
+  sv_rc = settings[_i].generate_key(NULL, &private_key, &private_key_size);
+  ck_assert_int_eq(sv_rc, SV_OK);
+
+  sv_rc = signed_video_set_private_key_new(sv, private_key, private_key_size);
+  ck_assert_int_eq(sv_rc, SV_OK);
+  sv_rc = signed_video_set_authenticity_level(sv, settings[_i].auth_level);
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:699]: Replace this line with setting the hash algo. That is relevant and is exercised in the test settings.

File: tests/check/check_h26xsigned_sign.c [hunk 29]
```diff
@@ -673,6 +673,49 @@ START_TEST(two_completed_seis_pending)
...
+  ck_assert_int_eq(sv_rc, SV_OK);
+  sv_rc = signed_video_generate_golden_sei(sv);
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:701]: Add empty line before to mark a clear separation to where we start generating the golden SEI.

File: tests/check/check_h26xsigned_sign.c [hunk 30]
```diff
@@ -978,6 +1021,8 @@ signed_video_suite(void)
   tcase_add_loop_test(tc, undefined_nalu_in_sequence, s, e);
   tcase_add_loop_test(tc, correct_timestamp, s, e);
   tcase_add_loop_test(tc, correct_signing_nalus_in_parts, s, e);
+  tcase_add_loop_test(tc, golden_sei_created, s, e);
+  tcase_add_loop_test(tc, golden_sei_created, s, e);
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:1025]: Remove one of these

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/signed_video_h26x_sign.c (hunk 12 above)
Comment (mithydolphin) [reply lib/src/signed_video_h26x_sign.c:238]: If statement gets too long if I do that

File: lib/src/includes/signed_video_sign.h [hunk 31]
```diff
@@ -336,16 +336,32 @@ signed_video_set_end_of_stream(signed_video_t *self);
  *       != SV_OK) {
  *     // Handle error
  *   }
- *
+ *   // Apply any configuration of choice.
  *   status = signed_video_generate_golden_sei(sv);
  *   if (status != SV_OK) {
  *     // Handle error
  *   }
+ *   // All necessary configurations need to be done prior to this call.
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:344]: You do not have to be this explicit. In fact, if the user really wants true code he/she can look in the test files for example. It is open-source. It is enough to say that "The user can get the golden SEI using signed_video_get_sei()."

Also, you should mention that the golden SEI in general is not generated instantly, hence the user has to wait for it. You can in the example code put a sleep(1) or something.

File: lib/src/includes/signed_video_sign.h [hunk 32]
```diff
@@ -317,6 +317,39 @@ signed_video_nalu_data_free(uint8_t *nalu_data);
...
+ *   }
+ *
+ *   status = signed_video_generate_golden_sei(sv);
+ *   if (status != SV_OK) {
+ *     // Handle error
+ *   }
+ *
+ * @param self Pointer to the signed_video_t object in use.
+ *
+ * @returns SV_OK            - Golden SEI is created,
```
[the remark this answers is not in the document]
Comment (bjornvolcker) [reply lib/src/includes/signed_video_sign.h:363]: I ment to replace both lines with this

File: lib/src/signed_video_h26x_sign.c [hunk 33]
```diff
@@ -221,11 +220,11 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
   SVI_TRY()
     // Get the total payload size of all TLVs. Then compute the total size of the SEI NALU to be
     // generated. Add extra space for potential emulation prevention bytes.
-    optional_document_size = tlv_list_encode_or_get_size(
-        self, optional_document_encoders, num_optional_document_encoders, NULL);
-    mandatory_document_size = !self->is_golden_sei
-        ? tlv_list_encode_or_get_size(self, mandatory_document_encoders, num_man_doc_encoders, NULL)
-        : 0;
+    optional_document_size =
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:223]: And since you're at it, maybe move away from "document" and use `optional_tags_size`
It is fine to rename `document_size` in this PR as well. Then use `tags_size`.

File: lib/src/signed_video_h26x_sign.c [hunk 34]
```diff
@@ -240,12 +239,9 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
       self->gop_info->list_idx = -1;  // Reset hash list size to exclude it from TLV
-      optional_document_size = tlv_list_encode_or_get_size(
-          self, optional_document_encoders, num_optional_document_encoders, NULL);
-      mandatory_document_size = !self->is_golden_sei
-          ? tlv_list_encode_or_get_size(
-                self, mandatory_document_encoders, num_man_doc_encoders, NULL)
-          : 0;
+      mandatory_document_size =
+          tlv_list_encode_or_get_size(self, mandatory_tags, num_mandatory_tags, NULL);
+      if (self->is_golden_sei) mandatory_document_size = 0;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:244]: I think it is better to check if `mandatory_document_size > 0` in the if on line 238, otherwise you'll simply enter and recalculate something that in the end again is set to zero.

File: lib/src/signed_video_h26x_sign.c (hunk 12 above)
Comment (bjornvolcker) [reply lib/src/signed_video_h26x_sign.c:238]: But it is better than entering here and then calculate the size again and after that set it to zero (again). Simply waste of operations.

File: lib/src/signed_video_h26x_sign.c [hunk 35]
```diff
@@ -304,13 +300,14 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
-              self, mandatory_document_encoders, num_man_doc_encoders, payload_ptr)
-        : 0;
+    size_t mandatory_written_size = 0;
+    if (mandatory_document_size > 0) {
+      mandatory_written_size =
+          tlv_list_encode_or_get_size(self, mandatory_tags, num_mandatory_tags, payload_ptr);
+    }
     size_t written_size = optional_written_size + mandatory_written_size;
     SVI_THROW_IF(written_size == 0, SVI_MEMORY);
     payload_ptr += mandatory_written_size;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:313]: Moving the pointer should be done right after the write operation, that is right after line 309.

File: lib/src/signed_video_tlv.c (hunk 21 above)
Comment (bjornvolcker) [reply lib/src/signed_video_tlv.c]: You forgot to make the name change in the header file

File: lib/src/signed_video_tlv.c (hunk 22 above)
Comment (bjornvolcker) [reply lib/src/signed_video_tlv.c]: You forgot to make the name change in the header file

File: tests/check/check_h26xsigned_sign.c [hunk 36]
```diff
@@ -674,18 +674,14 @@ START_TEST(two_completed_seis_pending)
 END_TEST
 
 /* Test description
- * Verify that after golden SEI is generated
- * The operation is as follows:
- * 1. Setup a signed_video_t session
- * 2. Generate a golden SEI
- * 3. Verify the golden SEI
+ * Generates a golden SEI and fetches it from the library.
+ * Then verifies that the correspond flag is set.
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:678]: nit: corresponding

File: tests/check/check_h26xsigned_sign.c [hunk 37]
```diff
@@ -696,10 +692,12 @@ START_TEST(golden_sei_created)
 
   sv_rc = signed_video_set_private_key_new(sv, private_key, private_key_size);
   ck_assert_int_eq(sv_rc, SV_OK);
-  sv_rc = signed_video_set_authenticity_level(sv, settings[_i].auth_level);
+  sv_rc = signed_video_set_hash_algo(sv, "sha256");
```
Comment (bjornvolcker) [inline tests/check/check_h26xsigned_sign.c:695]: Use `settings[_i].hash_algo_name`

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/includes/signed_video_sign.h [hunk 38]
```diff
@@ -342,26 +342,13 @@ signed_video_set_end_of_stream(signed_video_t *self);
  *     // Handle error
  *   }
  *   // All necessary configurations need to be done prior to this call.
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:344]: Remove this line. It is wrong. The necessary configuration has to be done before `signed_video_generate_golden_sei()`

File: lib/src/includes/signed_video_sign.h [hunk 39]
```diff
@@ -342,26 +342,13 @@ signed_video_set_end_of_stream(signed_video_t *self);
...
- *   status = signed_video_get_sei(sv, sei, &sei_size);
- *   if (status != SV_OK) {
- *     // Handle error
- *   }
- *   // The user is responsible for freeing |sei|.
- *   status = signed_video_get_sei(sv, NULL, &sei_size);
- *   if (status != SV_OK) {
- *       // True error. Handle it properly.
- *   }
+ *   // Before fetching the SEI user needs to wait for the SEI generated.s
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:345]: Remove "s" at end

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

File: lib/src/includes/signed_video_sign.h [hunk 40]
```diff
@@ -337,12 +337,12 @@ signed_video_set_end_of_stream(signed_video_t *self);
  *     // Handle error
  *   }
  *   // Apply any configuration of choice.
+ *   // All necessary configurations need to be done prior to this call.
```
Comment (bjornvolcker) [inline lib/src/includes/signed_video_sign.h:340]: Now you have two lines saying roughly the same thing. I think you can remove the second one.

File: lib/src/signed_video_h26x_sign.c [hunk 41]
```diff
@@ -220,29 +220,28 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
-    document_size = optional_document_size + mandatory_document_size;
+    document_size = optional_tags_size + mandatory_tags_size;
     payload_size = document_size + gop_info_size + vendor_size;
     payload_size += UUID_LEN;  // UUID
     payload_size += 1;  // One byte for reserved data.
-    if ((self->max_sei_payload_size > 0) && (payload_size > self->max_sei_payload_size)) {
+    if ((self->max_sei_payload_size > 0) && (payload_size > self->max_sei_payload_size) &&
+        (mandatory_tags_size > 0)) {
       // Fallback to GOP-level signing
       payload_size -= document_size;
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:240]: It should be enough to subtract `mandatory_tags_size` now.

File: lib/src/signed_video_h26x_sign.c [hunk 42]
```diff
@@ -304,13 +303,13 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
     payload_ptr += optional_written_size;
     size_t mandatory_written_size = 0;
-    if (mandatory_document_size > 0) {
+    if (mandatory_tags_size > 0) {
       mandatory_written_size =
           tlv_list_encode_or_get_size(self, mandatory_tags, num_mandatory_tags, payload_ptr);
+      payload_ptr += mandatory_written_size;
     }
     size_t written_size = optional_written_size + mandatory_written_size;
     SVI_THROW_IF(written_size == 0, SVI_MEMORY);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:312]: This will not catch the error if we run out of memory when writing the mandatory tags.

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
There are still some unresolved review comments from before

File: lib/src/signed_video_h26x_sign.c [hunk 43]
```diff
@@ -237,7 +237,7 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
     if ((self->max_sei_payload_size > 0) && (payload_size > self->max_sei_payload_size) &&
         (mandatory_tags_size > 0)) {
       // Fallback to GOP-level signing
-      payload_size -= document_size;
+      payload_size -= mandatory_tags_size;
       self->gop_info->list_idx = -1;  // Reset hash list size to exclude it from TLV
       mandatory_tags_size =
           tlv_list_encode_or_get_size(self, mandatory_tags, num_mandatory_tags, NULL);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:243]: On the line below, you are now adding the the `optional_tags_size`. This will make the payload_size incorrect.

File: lib/src/signed_video_h26x_sign.c [hunk 44]
```diff
@@ -299,16 +299,15 @@ generate_sei_nalu(signed_video_t *self, uint8_t **payload, uint8_t **payload_sig
...
+    SVI_THROW_IF(written_size == 0, SVI_MEMORY);
     if (mandatory_tags_size > 0) {
-      mandatory_written_size =
+      written_size =
           tlv_list_encode_or_get_size(self, mandatory_tags, num_mandatory_tags, payload_ptr);
-      payload_ptr += mandatory_written_size;
+      payload_ptr += written_size;
     }
-    size_t written_size = optional_written_size + mandatory_written_size;
     SVI_THROW_IF(written_size == 0, SVI_MEMORY);
```
Comment (bjornvolcker) [inline lib/src/signed_video_h26x_sign.c:311]: Put this inside the if-statement above

**Review by bjornvolcker — CHANGES_REQUESTED** [review]

**Review by bjornvolcker — CHANGES_REQUESTED** [review]
